### DISCUSSION TITLE
如何通过代码判断一个项目是否是umi脚手架搭建的项目

### DISCUSSION THREAD

**Discussion by zFitness** [umijs/umi#12054] [General]
我现在正在开发一款工具，需要判断一个前端工程是否是 umi 的项目，同时解析出项目中哪些文件为路由文件以及对应的路由信息，请教一下各位大佬。

**Comment by fz6m** [comment]
解析一下 `package.json` ，看看有没有 `umi` 或 `@umijs/max` 的依赖，如果是 `umi` ，可能有 `"umi": "2.x" | "3.x" | "4.x"` 的可能，如果是 `@umijs/max` ，只可能是 `"@umijs/max": "4.x"` ，这可以判断是不是 umi 的项目。

解析路由是不现实的，因为路由表是可以配置的，配置在配置文件的 `routes` 里面，而这些路由用到的哪些文件路径，完全是用户自己指定的，他们的位置不容易简单获取到，可以尝试把配置文件读取到文本（可能是 `.umirc.ts` 、`.umirc.${...}ts` 、`config.config.ts` 、`config/config.${...}.ts` 等，详见 [加载配置文件的优先级](https://umijs.org/docs/guides/env-variables#umi_env) ），之后把里面从 `umi` 或 `@umijs/max` 的 `defineConfig` 导入给剥离掉（ast 操作），之后 require string 得到配置的 object ，当然如果有心力的话还可以直接 ast 匹配出来 routes ，之后再从 routes 里面慢慢找到用到了哪些文件，总之这不是易事。

**Reply by zFitness** [reply]
umi 的配置文件可以用ts编写，好奇问一下是在运行Umi的过程中是如何把配置文件加载进来呢？

**Reply by fz6m** [reply]
umi 4 现在是 hack 了 require ，在导入过程中把 ts 转译成 cjs 的 js 再使用，但社区的做法各种各样，有用 babel 的，有用 native esm 的，require 过程中为了兼容 native esm 和 cjs 也有一些处理，总之我认为是比较复杂的，如果自己需要一种用 ts 做配置文件的工具，推荐直接用 [c12](https://github.com/unjs/c12) 吧，这也是 nuxt 中读 ts 配置文件用的库。
